[WIP] Change artifact behavior to moving instead of symlinking

mwean opened the pull request:
This is a first step towards change in installation behavior mentioned in #13201.
I had to skip the `two_apps_correct` tests because I don't know the desired functionality (e.g. should we copy the source to both targets and delete the source?)

RomainLanz commented:
Any news on this?

mwean commented:
I'm waiting for feedback from @jawshooah to make sure this is the correct approach.

jawshooah commented on lib/hbc/artifact/moved.rb:
```diff
@@ -0,0 +1,142 @@
+    self.class.artifact_english_name
+  end
+
+  def self.link_type_english_name
```
This method can be removed, as it's not used anywhere.

jawshooah commented on lib/hbc/artifact/moved.rb:
```diff
@@ -0,0 +1,142 @@
+    each_artifact do |artifact|
+      load_specification(artifact)
+      next unless preflight_checks
+      add_altname_metadata
```
I'm not sure how useful this is. Could you provide some justification for it?

jawshooah replied:
Didn't realize this was originally implemented in `Hbc::Artifact::Symlinked` and introduced in #3073 to address #2847. Judging from [this comment](https://github.com/caskroom/homebrew-cask/issues/2847#issuecomment-34649359), this shouldn't be necessary for actual files (i.e. not symlinks).

jawshooah commented on lib/hbc/artifact/moved.rb:
```diff
@@ -0,0 +1,142 @@
+  def existing_altname_metadata
+    result = @command.run(
+      '/usr/bin/xattr',
+      :args => ['-p', ALT_NAME_ATTRIBUTE, target],
```
Unless I'm mistaken, since this is called before `move`, there should be nothing at `target`. Where would the altnames come from?

vitorgalvao commented:
@mwean Just to check, do you still intend on working on this?

mwean commented:
@vitorgalvao Yeah I'd definitely like to. I just moved, but I'll try to take a look this weekend.

mwean commented:
@vitorgalvao I rebased against master and made a couple more updates. I'm not sure what else needs to be done, and I don't know what the correct behavior should be for the two-app tests. I could maybe copy to both locations and delete the original? I'm not sure what the use case for that is. Let me know what you think.

adidalal commented:
I agree that the two-app tests seem useless; at least, I've never seen it.
@jawshooah if you want to take a look at the PR when you're free?

jawshooah commented:
Will take another look at this tonight.

ptb commented:
I found one issue with this pull request, but unsure how to correctly correct it. If an app, Atom for example, has a binary that needs to be symlinked to /usr/local/bin, it complains that the symlink source isn't there. The workaround I've used is:

`sed -i -e "s/@cask.staged_path/Hbc.appdir/" /usr/local/Library/Taps/caskroom/homebrew-cask/lib/hbc/artifact/symlinked.rb`

but that breaks installs for fonts (at least). I'm not sure what the correct solution is, but figured I'd add that bit.

mwean commented:
@ptb are you saying that it tries to symlink the binary before it moves the app?

ptb commented:
No it tries to symlink after moving, but it is expects that it still lives at the staged location (Caskroom).

```
brew cask install atom --force
==> Satisfying dependencies
complete
==> Downloading https://github.com/atom/atom/releases/download/v1.4.3/atom-mac.zip
######################################################################## 100.0%
==> Verifying checksum for Cask atom
==> Moving App 'Atom.app' to '/Users/admin/Applications/Atom.app'
Error: It seems the symlink source is not there: '/opt/homebrew-cask/Caskroom/atom/1.4.3/Atom.app/Contents/Resources/app/apm/node_modules/.bin/apm'
```

vitorgalvao commented:
It had occurred to me recently this might be an issue. Basically this only happens if the binary to be linked is inside the `app`.

The solution would be to instead of having (to borrow `atom` as the example)

```ruby
binary 'Atom.app/Contents/Resources/app/apm/node_modules/.bin/apm', target: 'apm'
```

have

```ruby
binary "#{Hbc.appdir}/#{app[0]}/Contents/Resources/app/apm/node_modules/.bin/apm", target: 'apm'
```

Though that particular solution doesn’t work. All that’s needed is to add a new `app_location` (or something) special variable that we can grab in these cases, to just grab the install location.

mwean commented:
@ptb @vitorgalvao I made a failing spec for the embedded binary case, but I'm not sure how to proceed. I'm a little concerned putting interpolations into the path declaration like @vitorgalvao suggested because it would require Cask authors to know more about the cask inner workings. I wonder if we could match the beginning of the binary path against the app path (the "Atom.app" part in this example):

```
app 'Atom.app'
binary 'Atom.app/Contents/Resources/app/atom.sh', target: 'atom'
```

If there's a match, we'd know to look in the `appdir`. It would be a little ugly because we'd have to look at the app's artifacts from the symlink class, but it wouldn't require any change to the dsl. What do you think?

vitorgalvao commented:
I feel like you’re trying to make the feature smart, and smart is what got us in this mess in the first place [are we’re trying to avoid](https://github.com/caskroom/homebrew-cask/issues/13201).

It is not a huge user-facing change. I’d much rather have it be explicit, even if it means contributors need to know something new and working, than having something smart that later on we find breaks on some cases.

It’s better to have it be dumb and make it smarter later, than the reverse.

toonetown commented:
> @mwean
> it would require Cask authors to know more about the cask inner workings

I tend to agree with @vitorgalvao in that cask authors *should* be expected to know more about the cask inner workings.  It should be simple and dumb and easy for end users to use and not get into trouble...but  (IMO), it's OK to expect cask authors to need to know what they are doing.

It should be generally easy to create a cask...but it doesn't hurt to know the inner workings a bit in order to create one.

mwean commented:
Ok, that sounds good. What about having tokens instead of direct interpolation (like they do in [Paperclip](https://github.com/thoughtbot/paperclip/wiki/Interpolations))? Something like

```
binary ":install_location/:app/Contents/Resources/app/apm/node_modules/.bin/apm"
```

vitorgalvao commented:
Tokens seem clunky, to me. We use interpolation for some things, and symbols for others. They are well define and predictable. Mixing here seems inappropriate.

Also, we really need a way to specify the index of what we mean. Some apps have multiple `app` stanzas, so we need to specify the one we’re pointing to. Something like `binary "#{app_install_location}/#{app[0]}/…"` seems acceptable, to me. We need to specify it’s the install location reserved for `app`s, in case we later need to do it for other artifacts.

mwean commented:
Ok, I actually wasn't clear what `app[0]` was referring to in your earlier comment. Could we do something more explicit like `app_specifications[0].name`? Something shorter would be better but I can't think of anything right now.

vitorgalvao commented:
> Could we do something more explicit like `app_specifications[0].name`?

Not sure I get why. It seems more confusing to me. `app[0]` would reduce complexity, as we’re talking about the literal value of the `app` stanza. To clarify, let’s use a made up cask:

```ruby
app 'One.app'
app 'Two.app'
binary 'Two.app/Contents/Resources/app/script'
```

In this case, we’d use

```ruby
app 'One.app'
app 'Two.app'
binary "#{app_install_location}/#{app[1]}/Contents/Resources/app/script'"
```

I also wouldn’t oppose to just `appdir` instead of `app_install_location`.

To note that I’m not opposing `app_specifications[0].name`, I just don’t (yet?) see how it’s clearer than `app[0]`, seeing as it directly refers to the actual stanza.

mwean commented:
When I saw `app[0]`, I wasn't thinking of having multiple apps, I was thinking of args to the `app` method or something, since we destructure the args elsewhere (e.g. `source_string, target_hash = artifact_spec`). At any rate, if we're saying `app` is the collection of app stanzas, then why would `app[0]` necessarily return the app name? I'm totally fine either way, just want to make sure it's clear as possible.

vitorgalvao commented:
Right. From a user/contributor-facing view, I was thinking of `app` as “a string that indicates what will be acted upon”.

Lets look at it another way, though. What would be easiest/would make more sense for you to implement?

mwean commented:
I'll try to spike on it a bit tonight and let you know.

vitorgalvao commented:
Sounds good. Thank you.

mwean commented:
Ok, I added the `appdir` method to the DSL for interpolation. I decided to leave out the `app[0]` part because current casks (e.g. Atom) are already duplicating the app ("Atom.app") part, and it's much simpler to just continue with that. @ptb, this should now work with the embedded binaries. Let me know what you think.

vitorgalvao commented:
> current casks (e.g. Atom) are already duplicating the app ("Atom.app") part, and it's much simpler to just continue with that.

Fair enough.

> @ptb, this should now work with the embedded binaries.

But you’ll still need to update the casks by hand first, to try it out.

@caskroom/maintainers We should test this for a bit and see if we find any issue while using.

adidalal commented:
What's the best way to test locally?

mwean commented:
@vitorgalvao should I work on a script or something to update all the existing casks with apps and binaries? That seems like it would take some work.

vitorgalvao commented:
Best way to test locally:

```bash
git clone https://github.com/mwean/homebrew-cask.git # clone @mwean’s repo with this change
./homebrew-cask/developer/bin/develop_brew_cask # make his repo the main one to use
```

As `develop_brew_cask` warns, do not run `brew update` while in this mode. To get back to regular usage, run `production_brew_cask` in that same directory.

@mwean Just tested on a clean VM, and noted something. This is still using `~/Applications` by default. With this change, `/Applications` should be the default. With that, only point **5** of the “The end of linking, and the beginning of moving (technical)” section in #13201 will remain of the first part of this change.

As for the script, you may work on it if you want, but if not, I’d be happy to do it. It’s likely many of the casks will have to be manually altered anyway, and that’s a boring job that I’m prepared to do.

jawshooah commented:
Note that, if you already have a local clone of the repo, you can do the following from its root:

```
brew install hub
hub checkout https://github.com/caskroom/homebrew-cask/pull/13966
./developer/bin/develop_brew_cask
```

mwean commented:
@vitorgalvao this only handles `app` stanzas now. Should I switch all the other artifacts (except binary) to use move?

I'll update the default `appdir`. 

I'll try looking at writing a migration script this weekend, but feel free to tackle it if you have time before then.

vitorgalvao commented:
> Should I switch all the other artifacts (except binary) to use move?

Yes, might as well switch them all from the start.

mwean commented:
I forgot that I skipped the two-app specs because I didn't know how to handle that case. We could copy the app and then do some sort of cleanup at the end of the cask installation or something. I don't really understand the use-case anyway.

mikemcquaid commented:
@vitorgalvao Something we've found useful for testing relatively invasive changes like this is guarding the new behaviour behind an environment variable (e.g. `CASK_MOVE_APPS`), disabling it by default and then allowing people to enable it gradually. That also means you can merge PRs before needing to ship it to people so review becomes easier.

claui commented:
I ❤️  this PR and the way it is implemented!

@mwean How are we going to handle `brew cask install --force`?
With plain `brew cask install`, I still see the warning:

> Add the "--force" option to force re-install.

… which sets up my expectation that adding `--force` would trash and replace the existing app (if and only if the app is managed by brew-cask).

What I’m observing instead is that `brew cask install --force` now [refuses to overwrite the existing app](https://asciinema.org/a/1ai31r2y5ascgtk4bx2x9hshx). I suggest this to be changed to avoid confusion and to give the `--force` option a chance to continue making sense.

Any thoughts?

mwean commented:
@claui that's a good point. I was also thinking about that when I was looking at the specs that prevent clobbering existing artifacts. I was thinking that when we handle updates, it would be hard to be sure that we're overwriting the app that we installed and not one the user subsequently installed manually. Personally, I'd be happy to have it always overwrite, but a maintainer should weigh in on the desired behavior.

adidalal commented:
Agreed with the always overwrite - it *shouldn't* matter where the app came from, as desired user behavior should result in the latest (version in Cask) should be installed....

but in fact, this is where `auto_updates true` will be of use, as maybe that should be the only reason to not overwrite the existing app (if I'm remembering the original proposal correctly). We can handle misbehaving/auto-updating Casks on a manual basis after the initial merge, as adding that line isn't particularly onerous (and even probably scriptable, to be honest)

vitorgalvao commented:
@mwean The use case is we have some casks that really do have two `app` stanzas. Some packages really consist of multiple `app` bundles.

Regarding `auto_updates`, the rule is: when we have `upgrade` functionality, every installed cask that has `auto_updates` (even if the cask was updated) will be ignored by us (and we’ll let the app take care of the update).

As for overwriting, I think it’s good we **do not** overwrite by default, but **do** overwrite when using `--force`. The reason is simple: it’s not at all uncommon for different casks to have artifacts with the same name (see all `eclipse` casks). Overwriting those by default could be a nasty surprise.

@MikeMcQuaid That is a great idea, and we should adopt that as well. For this particular case, however, I’m fine with letting the breaking change come in. Reason being we’re so low on core developers, this important feature is being contributed by someone outside the main team, and I really do not want to take @mwean’s time more than necessary. We’ve also [warned everyone things would break](https://github.com/caskroom/homebrew-cask/issues/13201) and the response has been quite positive (people really want this new behaviour), so I’m fine with it.

mwean commented:
> The use case is we have some casks that really do have two app stanzas. Some packages really consist of multiple app bundles.

That makes sense, but the spec is specifically for the same app going to two different locations:

```ruby
# test/support/Casks/with-two-apps-correct.rb
app 'Caffeine.app'
app 'Caffeine.app', target: 'AnotherName.app'
```

Having two app stanzas is fine, as long as they don't use the same source. Do you still want to support the use case in this spec?

> As for overwriting, I think it’s good we do not overwrite by default, but do overwrite when using --force. 

Sounds good. This should work correctly with @claui's branch.

claui commented:
@mwean I have checked every cask with two or more `app` stanzas:

```
grep -c '^\s*app\s' Casks/*.rb | ruby -ne 'puts $1 if ~/(.*):[^01]/' | xargs mate
```

and found that the scenario from the spec never comes up even once. I would suggest that that particular spec be abandoned because it has no redeeming value.

claui commented:
> This should work correctly with @claui's branch.

Speaking of working correctly, I’ve been bitten earlier today with one more issue: `brew cask install lingo --force` finds an existing target, tries to trash it but fails because I happen not to be the owner.

Why not? Turns out I had bought the same app from the MAS years ago but forgot about it. Today I had my head in the clouds as always so I ran `install lingo --force` on my branch. The conflicting Lingo.app, being a MAS app, belonged to root so it didn’t budge. I feel it’s pretty safe to assume that in general, **end&nbsp;users will not remember** whether their app came from the MAS or HBC.

So I’m left wondering what HBC should do in that case?

1. Leave the artifact staged and display the *Not moved* message?
- Go `sudo`, asking the user for the password?
- Give up and ask the user to manually resolve the conflict?

@mwean @vitorgalvao Any thoughts as to what would be the most reasonable course of action here?

vitorgalvao commented:
@mwean Yes, the way that spec is presented is a bit weird. There are indeed some casks with more than one `app`, but no, never the same, as @claui pointed out.

All that said, I agree with [your previous comment](https://github.com/caskroom/homebrew-cask/pull/13966#issuecomment-219922983), lets leave out the whole `app[0]` nonsense. It works well as is.

@claui Your second scenario seems the most reasonable. If we’re saying `--force`, then forceful it should be, consequences be damned (presumably, the user understands that `force` is the “do whatever it takes to make this work” option).

That said, I have no strong commitment to that behaviour. If anyone thinks one of the others should be the case and makes a good case for it, I’m fine either way.

mwean commented:
Great, I'll update the 2-app specs and merge in @claui's branch.

mwean commented:
@vitorgalvao @claui: Ok, I fixed the 2-app spec, updated all the binary stanzas, and merged in @claui's branch. I'm not sure what the failure is on Travis, but I think we might be good to go with this!

adidalal commented:
@mwean Ignore the Travis error - it's been nearly impossible to pin down, but it's in no way indicative of an actual failure/error in the PR.

Will let one of the other maintainers have a look, but LGTM.

Easiest way to "migrate" might just be to get a list of all casks, uninstall them and wipe out /Caskroom and associated symlinks, and then reinstall after this change is merged, but open to better methods - I wouldn't be opposed to tagging a new release as well, just as a symbolic measure.

muescha commented:
i would prefer moving apps out of /Caskroom is better than reinstalling app (maybe someone dont like a new version installed)

claui commented:
@adidalal Are you thinking of a scripted migration? Would such a migration be optional or rather mandatory?

adidalal commented:
No opinion either way - with such a large change, there probably isn't a "best" way to proceed, but open to ideas for how to make it a not-terrible experience. If it's getting off-topic here, this discussion can be moved to a new issue, but it's one that should happen before this change is merged and made final

vitorgalvao commented:
@muescha @claui @adidalal No migration. That’s the point made at the end of https://github.com/caskroom/homebrew-cask/issues/13201.

> #### Doesn’t this mean a bunch of stuff might break?

> (…) transition strategies is an inglorious job (everything will be discarded in the end, after all) that puts on hold [even the smallest of changes](https://github.com/caskroom/homebrew-cask/pull/9178) indefinitely. It is incredibly time consuming, has variable results (we still get a bunch of issues pertaining to it), and (above all) kills momentum. Furthermore, this is a big change and there’s really no good way to make this seamlessly.

If anyone wants to devise a transition strategy, that is fine. We’d need commitment to that, however, meaning the transitioner will be responsible for implementation, bugs pertaining to it, and removing the code in a few months when everyone is presumably transitioned.

#13201 made clear from the start we’re willing to break things with this change. All things considered, it’s not even a huge break, your old installs won’t stop working. **I’ll greatly advise anyone against a transition strategy for this**. Even [`brew cask list` is broken](https://github.com/caskroom/homebrew-cask/issues/14058) and we don’t keep track of installed casks correctly. There’s no way you’ll write a good transition when everything else is falling around you and you can’t be sure you can lean on other parts of the codebase.

Lets leave it be. Accept the breakage as we already knew it would happen, and move on.

Leaving this open for a few more days for disagreements, but I’ll say I’m excited to have this merged in. Not everything in #13201 is present, but the most important bits are, and it’s another step to the finally right direction.

ptb commented:
I believe the changes to all of the casks to make "#{appdir}" work correctly will need to be enclosed in double quotes, otherwise these are the types of errors that will be produced:

`Error: It seems the symlink source is not there: '/usr/local/Caskroom/atom/1.7.3/#{appdir}/Atom.app/Contents/Resources/app/apm/node_modules/.bin/apm`

mwean commented:
Great catch @ptb, I'll get that fixed

mwean commented:
Ok, should be fixed now

ptb commented:
The casks that specify a target for the binary didn't get changed (e.g. Atom, Github Desktop, others).
_(And I'll say **thank you** now, for all of your work on this!)_

jawshooah commented on Casks/atom.rb:
```diff
@@ -14,8 +14,8 @@
   app 'Atom.app'
-  binary 'Atom.app/Contents/Resources/app/apm/node_modules/.bin/apm', target: 'apm'
-  binary 'Atom.app/Contents/Resources/app/atom.sh', target: 'atom'
+  binary '#{appdir}/Atom.app/Contents/Resources/app/apm/node_modules/.bin/apm', target: 'apm'
```
Still using single-quotes here, and in other casks.

mwean commented:
Ah, gotta fix my regex...

mwean commented:
Ok, let's try that again

jawshooah commented:
Need to change `dest_path` to `Hbc.appdir` [here](https://github.com/caskroom/homebrew-cask/pull/13966/files#diff-2f4bc7a44ed551d4bf50c3ab80068f90R113) as well.

mwean commented:
I'm trying to reproduce the failure on CI, but I'm getting `Hbc::CaskError: Uh oh, could not figure out how to unpack '/var/folders/81/krskm0797s78hqmfdx7v4_k40000gn/T/homebrew_tests20160527-29970-m1ln87/cache/stuffit-container-1.2.3.sit'`. Could someone help me with this?

jawshooah commented:
@mwean Do you have `unar` installed?

Separate but related issue: that test uses a test cask containing an `artifact` stanza with no `target`, which should never happen. Will fix in a bit.

mwean commented:
Haha I was reading that as un**r**ar... Problem solved.

@jawshooah I added `target: sheldonmac/v1` and it's working now, thanks!

jawshooah commented:
@mwean There are some conflicts now that #21549 has been merged in. Resolve those and push, and I'll be glad to merge this once Travis goes green.

jawshooah commented:
@caskroom/maintainers
![Whaaaaaaaaat](https://media.giphy.com/media/rl0FOxdz7CcxO/giphy.gif)

mwean commented:
@jawshooah I think you're good to go!

claui commented:
Yay!!! What a fantastic day for Homebrew Cask! <3

mwean commented:
So what's the next step @vitorgalvao??

vitorgalvao commented:
Next steps:

+ Get rid of versioned subdirectories in the Caskroom. The `.metadata` directory will still need to exist (although its innards won’t have the versioned app directory either) since it will be useful to save a snapshot of the installed cask to handle updates/uninstalls at a future time.
+ Implement `:required_location` (apps that will benefit from it at the start are mentioned in #12858).
    + And with it, `--force-appdir` (these two really go hand in hand).

And that is actually way less of #13201 left than I realised!

We still have a lot to do after that, but closing #13201 will be a huge milestone (this merged pull request already is!), and it makes me really glad to see the project completing this step, one further in the right direction.

Thank you greatly, @mwean, @claui, @jawshooah, and many others (too many to mention all without missing some and making this sound like an award ceremonies speech) for all the work.

muescha commented on USAGE.md:
```diff
@@ -176,35 +176,34 @@ You can also modify the default installation locations used when issuing `brew c
-# Will force the Chrome app to be linked to ~/Applications
-# even though HOMEBREW_CASK_OPTS specified /Applications
-$ brew cask install --appdir="~/Applications" google-chrome
+# Will force the Chrome app to be linked to /Applications
```
`linked`? 

better: `created in` or `moved to`

muescha commented on doc/cask_language_reference/all_stanzas.md:
```diff
@@ -33,7 +33,7 @@ Each Cask must declare one or more *artifacts* (i.e. something to install).
 | `vst_plugin`       | yes                           | relative path to a VST Plugin that should be linked into the `~/Library/Audio/VST` folder on installation
 | `vst3_plugin`      | yes                           | relative path to a VST3 Plugin that should be linked into the `~/Library/Audio/VST3` folder on installation
-| `suite`            | yes                           | relative path to a containing directory that should be linked into the `~/Applications` folder on installation (see also [Suite Stanza Details](stanzas/suite.md))
+| `suite`            | yes                           | relative path to a containing directory that should be linked into the `/Applications` folder on installation (see also [Suite Stanza Details](stanzas/suite.md))
```
`linked` --> `moved`?

muescha commented on man/man1/brew-cask.1:
```diff
@@ -149,7 +149,7 @@ Give additional feedback during installation\.
 .TP
 \fB\-\-appdir=<path>\fR
-Target location for Application links\. The default value is \fB~/Applications\fR\.
+Target location for Application links\. The default value is \fB/Applications\fR\.
```
it is not a link anymore

muescha commented on USAGE.md:
```diff
@@ -176,35 +176,34 @@ You can also modify the default installation locations used when issuing `brew c
 * `--caskroom=/my/path` determines where the actual applications will be located.
 Should be handled with care — setting it outside `/opt` or your home directory might mess up your system.
 Default is `/opt/homebrew-cask/Caskroom`
```
confused with `caskroom` and `appdir` -> we still need `caskroom`?

vitorgalvao replied:
Yes, we still need to Caskroom. See #13201, specifically point 4 of this screenshot:
![image](https://cloud.githubusercontent.com/assets/1699443/15708484/a4695974-27f6-11e6-871f-fea54b0e9be3.png)

muescha replied:
(#13201 is closed for comments - so i answer here)

so it is a "cache" where are placed "leftovers" from installation? i expected that hbc creates only temp dirs and then deleted all other stuff.

it should be better described in README/USAGE and whats the difference.

vitorgalvao replied:
>  i expected that hbc creates only temp dirs and then deleted all other stuff

That isn’t temporary stuff, though. Caskroom will keep three things (one of which is not mentioned there):

+ A `.metadata` directory that contains a snapshot of the state of the cask in the moment you installed it (this will be useful in the future to correctly deal with uninstalls)
+ Any CLI artifacts that will be linked to `/usr/local/bin/` (or whatever). Since they are links, they need a place to be where they are linked from.
+ Uninstall scripts and other things that come in the `dmg` (or whatever) that although not integral to the app, might be relevant t to some users.

muescha replied:
thx for details :)

this info should be added to the README/USAGE

vitorgalvao replied:
Since the `google-chrome staged at '/opt/homebrew-cask/Caskroom/google-chrome/latest'` message is already output, a [FAQ](https://github.com/caskroom/homebrew-cask/tree/master/doc/faq) document might be more appropriate, for inquiring users (I’ll wager most won’t care, so having this in such a prominent place like the README stops making sense).

muescha replied:
ah - ok - i dont see the line `google-chrome staged at ...` before. it makes is more clear

but since its moved:

if an artefact has only one binary/app then this staged dir in `Caskroom` leaves empty (i assume the `.metadata` is per app but not always visible)?

toonetown replied:
Are there plans to perhaps move the caskroom directory under `$(brew --prefix)` at some point now that artifacts are moved instead of linked?  I think that the original reason for putting it in `/opt` is that spotlight would index there - but it wouldn't index in `/usr/local` (https://github.com/caskroom/homebrew-cask/issues/2494)...but now that artifacts are actually moved, it wouldn't hurt (IMO) to put the Caskroom by *default* at `$(brew --prefix)/Caskroom`.

vitorgalvao replied:
@muescha Yes, the directory may *look*  empty (it will have `.metadata`), and the `staged at` message will even say `0B`. Not a huge issue to have it there, though.

@toonetown Yes, `/opt` was originally chosen for that reason, but later on it proved to be the [right directory anyway](http://www.pathname.com/fhs/pub/fhs-2.3.html#OPTADDONAPPLICATIONSOFTWAREPACKAGES).

All that said, since we’re more integrated with homebrew and that integration will only become bigger, I’d have no quarrel in moving under `$(brew --prefix)`. It will even save us from having to ask for a password on first install.

Mind opening a new issue so we can discuss it?

muescha commented on README.md:
```diff
@@ -30,7 +30,7 @@ $ brew cask install google-chrome
 And there we have it. Google Chrome installed with a few quick commands: no clicking, no dragging, no dropping.
```
code snippet for `brew cask install google-chrome` should be changed from `Symlinking` to `Moving` (line 23 of collapsed code in diff)

vitorgalvao replied:
This one I have already fixed yesterday.

muescha commented on USAGE.md:
````diff
@@ -53,7 +53,7 @@ Easy peasy:
 $ brew cask uninstall google-chrome
 ```
````
code snippet for `brew cask install google-chrome` should be changed from `Symlinking` to `Moving` (line 44 of collapsed code in diff)

vitorgalvao replied:
This one I have already fixed yesterday as well.

mwean commented:
@muescha I think there was a bad rebase that reverted my doc changes. I'll make a new PR to fix them

vitorgalvao commented:
Fixing issues caught by @muescha in https://github.com/caskroom/homebrew-cask/pull/21597.

toonetown commented:
I just have to say, it might be completely in my head (or due to the fact that I uninstalled/reinstalled all my casks to get the "move" behavior), but my applications seem to launch a *LOT* faster and spotlight seems to be a *LOT* more responsive  since this change has been implemented.  Nice work!!!
